### PR TITLE
common: stroke dash offset support added

### PR DESCRIPTION
**Pull request by mgrudzinska**
The feature is supported also in the svg loader.

@Issue: https://github.com/thorvg/thorvg/issues/1591

### CODE REVIEW

**Comment by mgrudzinska** [comment]
sample:
```
<svg viewBox="-10 0 50 55" xmlns="http://www.w3.org/2000/svg">
  <line x1="0" y1="1" x2="30" y2="1" stroke="black" />
  <line x1="0" y1="3" x2="30" y2="3" stroke="black" stroke-dasharray="3.4 1.3 2 0.5" stroke-linecap="butt" />

  <line x1="0" y1="5" x2="30" y2="5" stroke="blue"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="1.2" />
  <line x1="0" y1="6.1" x2="30" y2="6.1" stroke="blue"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="8.4" />

  <line x1="0" y1="8" x2="30" y2="8" stroke="blue"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="4.1" />
  <line x1="0" y1="9.1" x2="30" y2="9.1" stroke="blue"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="11.3" />

  <line x1="0" y1="11" x2="30" y2="11" stroke="blue"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="5.7" />
  <line x1="0" y1="12.1" x2="30" y2="12.1" stroke="blue"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="12.9" />


  <line x1="0" y1="14" x2="30" y2="14" stroke="red"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="-1.2" />
  <line x1="0" y1="15.1" x2="30" y2="15.1" stroke="red"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="-8.4" />

  <line x1="0" y1="17" x2="30" y2="17" stroke="red"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="-4.1" />
  <line x1="0" y1="18.1" x2="30" y2="18.1" stroke="red"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="-11.3" />

  <line x1="0" y1="20" x2="30" y2="20" stroke="red"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="-5.7" />
  <line x1="0" y1="21.1" x2="30" y2="21.1" stroke="red"
        stroke-dasharray="3.4 1.3 2 0.5" stroke-dashoffset="-12.9" />
  

  <line x1="0" y1="23" x2="30" y2="23" stroke="green"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="-1.2" />
  <line x1="0" y1="24.1" x2="30" y2="24.1" stroke="green"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="-8.4" />

  <line x1="0" y1="26" x2="30" y2="26" stroke="green"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="-4.1" />
  <line x1="0" y1="27.1" x2="30" y2="27.1" stroke="green"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="-11.3" />

  <line x1="0" y1="29" x2="30" y2="29" stroke="green"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="-5.7" />
  <line x1="0" y1="30.1" x2="30" y2="30.1" stroke="green"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="-12.9" />

  
  <line x1="0" y1="32" x2="30" y2="32" stroke="blue"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="1.2" />
  <line x1="0" y1="33.1" x2="30" y2="33.1" stroke="blue"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="8.4" />

  <line x1="0" y1="35" x2="30" y2="35" stroke="blue"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="4.1" />
  <line x1="0" y1="36.1" x2="30" y2="36.1" stroke="blue"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="11.3" />

  <line x1="0" y1="38" x2="30" y2="38" stroke="blue"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="5.7" />
  <line x1="0" y1="39.1" x2="30" y2="39.1" stroke="blue"
        stroke-dasharray="3.4 1.3 2" stroke-dashoffset="12.9" />

  
  <line x1="0" y1="41" x2="30" y2="41" stroke="black" stroke-dasharray="3.4 1.3 2"/>
</svg>
```

result:
Zrzut ekranu 2023-08-17 o 23 35 27

File: src/loaders/svg/tvgSvgLoaderCommon.h [hunk 1]
```diff
@@ -470,6 +472,7 @@ struct SvgStyleStroke
     float miterlimit;
     SvgDash dash;
     int dashCount;
+    float dashOffset;
```
Comment (hermet) [inline src/loaders/svg/tvgSvgLoaderCommon.h:475]:
```
SvgDash {
  - array
  - offset
}
```
dashCount??  ->  dash.array.count

File: src/lib/tvgRender.h [hunk 2]
```diff
@@ -200,10 +201,11 @@ struct RenderShape
         return stroke->fill;
     }
 
-    uint32_t strokeDash(const float** dashPattern) const
+    uint32_t strokeDash(const float** dashPattern, float* offset = nullptr) const
```
Comment (hermet) [inline src/lib/tvgRender.h:204]: nullptr assignment is not quite meaningful. remove it please.

**Review by hermet — COMMENTED** [review]
thanks. a nitpick. please check a comment.

**Review by hermet — CHANGES_REQUESTED** [review]
It is beneficial to set a default value at Shape::fill() . I believe that reverting it is not a good idea.

File: src/bindings/capi/thorvg_capi.h [hunk 3]
```diff
@@ -1339,6 +1339,24 @@ TVG_API Tvg_Result tvg_shape_get_stroke_gradient(const Tvg_Paint* paint, Tvg_Gra
 TVG_API Tvg_Result tvg_shape_set_stroke_dash(Tvg_Paint* paint, const float* dashPattern, uint32_t cnt);
```
Comment (hermet) [inline src/bindings/capi/thorvg_capi.h:1339]: Please kill the old version of the api.  I think we don't need to maintain both apis.

File: src/bindings/capi/thorvg_capi.h [hunk 4]
```diff
@@ -1336,23 +1330,24 @@ TVG_API Tvg_Result tvg_shape_get_stroke_gradient(const Tvg_Paint* paint, Tvg_Gra
 *
 * \note To reset the stroke dash pattern, pass @c nullptr to @p dashPattern and zero to @p cnt.
 */
-TVG_API Tvg_Result tvg_shape_set_stroke_dash(Tvg_Paint* paint, const float* dashPattern, uint32_t cnt);
+TVG_API Tvg_Result tvg_shape_set_stroke_dash(Tvg_Paint* paint, const float* dashPattern, uint32_t cnt, float offset);
 
 
 /*!
-* \brief Gets the dash pattern of the stroke.
```
Comment (mgrudzinska) [inline src/bindings/capi/thorvg_capi.h:1343]: @hermet I added BETA_API - not sure whether it's necessary.

File: inc/thorvg.h [hunk 5]
```diff
@@ -980,11 +980,11 @@ class TVG_API Shape final : public Paint
      * @param[in] r The red color channel value in the range [0 ~ 255]. The default value is 0.
      * @param[in] g The green color channel value in the range [0 ~ 255]. The default value is 0.
      * @param[in] b The blue color channel value in the range [0 ~ 255]. The default value is 0.
-     * @param[in] a The alpha channel value in the range [0 ~ 255], where 0 is completely transparent and 255 is opaque. The default value is 0.
+     * @param[in] a The alpha channel value in the range [0 ~ 255], where 0 is completely transparent and 255 is opaque.
      *
      * @return Result::Success when succeed, Result::FailedAllocation otherwise.
      */
-    Result stroke(uint8_t r, uint8_t g, uint8_t b, uint8_t a = 255) noexcept;
```
Comment (mgrudzinska) [inline inc/thorvg.h:987]: @hermet our policy is that we can remove the def value from api just like that?

File: src/bindings/capi/thorvg_capi.h (hunk 4 above)
Comment (hermet) [reply src/bindings/capi/thorvg_capi.h:1343]: If you think it's ready to official release then you don't need to add BETA_API  tag. Otherwise you should.

Comment (hermet) [reply src/bindings/capi/thorvg_capi.h:1343]: Ok, now this breaks the api compatibility.

**Review by hermet — APPROVED** [review]
great job!

**Pull request merged**